src/virtualprocess.jl

```diff
@@ -179,131 +185,157 @@ function virtual_process!(toplevelex::Expr,
-                if is_assignment(x′)
-                    # local a, b = sincos(1)
-                    vars = get_lhs(x′)
-                    push!(locally_scoped_vars, vars...)
-                else
-                    # local a, b ...
-                    push!(locally_scoped_vars, x.args...)
-                end
-                return x
+        # interpret type definitions
+        istypedef(stmt) && return true
+        # add more statements necessary for the first time interpretation of a type definition
+        # TODO: maybe upstream these into LoweredCodeUtils.jl ?
+        if isexpr(stmt, :(=))
+            stmt = stmt.args[2] # rhs
+        end
+        if isexpr(stmt, :call)
+            f = stmt.args[1]
+            if isa(f, GlobalRef)
+                f.mod === Core && f.name in (:_setsuper!, :_equiv_typedef, :_typebody!) && return true
             end
-
-            # annotate `global`s for regular assignments in global scope
-            if !(islocalscope(scope) || is_already_scoped(scope))
-                # `parse_input_line` will parse "a, b = sincos(1)" into `:((a, b) = sincos(1))`
-                if is_assignment(x)
-                    # FIXME: this can't handle following case
-                    # begin
-                    #     local l
-                    #     l, g = sincos(1) # won't be `global` annotated
-                    # end
-                    vars = get_lhs(x)
-                    any(in(locally_scoped_vars), vars) || return Expr(:global, x)
-                end
+            if isa(f, QuoteNode)
+                f.value in (Core._setsuper!, Core._equiv_typedef, Core._typebody!) && return true
             end
-
-            return x
         end
 
-        @debug x
+        return false
+    end
 
-        shouldprofile(x) || continue
+    lines_required!(not_abstracted, src, CodeEdges(src))
 
-        λ = gen_virtual_lambda(virtualmod, lnn, x)
+    return not_abstracted
+end
 
-        interp = TPInterpreter(; # world age gets updated to take in `λ`
-                               inf_params              = InferenceParams(interp),
-                               opt_params              = OptimizationParams(interp),
-                               optimize                = may_optimize(interp),
-                               compress                = may_compress(interp),
-                               discard_trees           = may_discard_trees(interp),
-                               filter_native_remarks   = interp.filter_native_remarks,
-                               )
+isinclude(ex) = isexpr(ex, :call) && first(ex.args) === :include
+
+# adapted from https://github.com/JuliaDebug/JuliaInterpreter.jl/blob/2f5f80034bc287a60fe77c4e3b5a49a087e38f8b/src/interpret.jl#L188-L199
+# works as `JuliaInterpreter.evaluate_call_compiled!`, but special cases for TypeProfiler.jl added
+function evaluate_call_recurse!(interp::ActualInterpreter, frame::Frame, call_expr::Expr; enter_generated::Bool=false)
```

Comment:
Today I noticed that defining this method triggers more than 500 invalidations if you've loaded Revise. (In contrast, extending `step_expr!` only triggered 8 invalidations, though if one gets fixed it might alter the count for the other.) We can live with that, but it might also be worth considering whether there are alternatives:
- define some inner call in JuliaInterpreter's `evaluate_call_recurse!` method that obviates the need for this extension (i.e., can we add a branch that handles `include`s?)
- move the definition of `ConcreteInterpreter` and this second method to JuliaInterpreter.jl (I suspect that is not a good solution)
- always use runtime dispatch to call `evaluate_call_recurse!`